src/programs.c

```diff
@@ -637,5 +778,1001 @@ int uprobe__read_return_string(struct pt_regs *ctx)
     return 0;
 }
 
+static __always_inline void clear_telemetry_events()
+{
+    // The key is always zero since the index is just a simple state counter
+    u32 key = 0;
+    bpf_map_update_elem(&telemetry_index, &key, &key, BPF_ANY);
+    bpf_map_update_elem(&read_flush_index, &key, &key, BPF_ANY);
+}
+
+static __always_inline void push_telemetry_event(ptelemetry_event_t ev)
+{
+    u32 key = 0;
+    u32 *pcurrent_index = bpf_map_lookup_elem(&telemetry_index, &key);
+    if (NULL == pcurrent_index)
+    {
+        return;
+    }
+    u32 idx = *pcurrent_index + 0;
+
+    int ret = bpf_map_update_elem(&telemetry_stack, &idx, ev, BPF_ANY);
+    if (ret < 0)
+    {
+        return;
+    }
+
+    // Update the index
+    // this is fine:
+    idx = *pcurrent_index + 1;
+    // // This fails verification on 4.15 with `invalid register spill`
+    // idx += 1;
+    // // this fails too, same message
+    // idx = idx + 1;
+    ret = bpf_map_update_elem(&telemetry_index, &key, &idx, BPF_ANY);
+    if (ret < 0 )
+    {
+        return;
+    }
+
+    ptelemetry_event_t pev = bpf_map_lookup_elem(&telemetry_stack, &idx);
+    if (pev)
+    {
+        __builtin_memcpy(ev, pev, sizeof(telemetry_event_t));
+    }
+    return;
+}
+
+#define FLUSH_LOOP                                                          \
+    if (ii < *pcurrent_index) {                                            \
+        ev = bpf_map_lookup_elem(&telemetry_stack, &ii);                     \
+        if (ev) {                                                           \
+            bpf_perf_event_output(ctx,                                      \
+                &telemetry_events,                                          \
+                bpf_get_smp_processor_id(),                                 \
+                ev,                                                         \
+                sizeof(*ev)                                                 \
+            );                                                              \
+            ii++;                                                        \
```

Comment:
don't we always want to increment ii?

Reply:
There _shouldn't_ be a case where we get a null pointer back from `bpf_map_lookup_elem` and there are events past it on the stack, so with `ii++` inside or outside the `if` we'll just keep skipping down the loop until we get to the end. It also shouldn't make a difference to move ie outside the `if (ev)` either and it would catch events if we accidentally skipped an element somehow.

Reply:
750d697f2242e69233ef10fb19251be28b8c7bae